### PR TITLE
[FlagGems Operator Development Competition] Add grid_sample operator

### PR DESCRIPTION
**Pull request by goldenfox2025**
## Summary

Implements `grid_sample` with Triton kernels and registers these ATen entries:

- `aten.grid_sampler`
- `aten.grid_sampler_2d`
- `aten.grid_sampler_3d`
- `aten.grid_sampler_2d_backward`
- `aten.grid_sampler_3d_backward`

Supported API surface matches `torch.nn.functional.grid_sample`:

| Dimension | Interpolation | Padding | align_corners |
|-----------|---------------|---------|---------------|
| 2D (4-D input) | bilinear, nearest, bicubic | zeros, border, reflection | True / False |
| 3D (5-D input) | trilinear, nearest | zeros, border, reflection | True / False |

`torch.nn.functional.grid_sample` backward is also wired through `torch.library.register_autograd` on `aten::grid_sampler`, so the unified functional path and the direct ATen path both match PyTorch behavior.

## Implementation

The kernels use a flattened spatial tiling strategy:

- 2D forward/backward tile `H_out x W_out`
- 3D forward/backward tile `D_out x H_out x W_out`
- Channel reduction stays inside each tile so grid coordinate transforms are computed once per output position and reused across channels

Backward uses separate kernels for `grad_input` and `grad_grid` because their access patterns differ:

- `grad_input` uses scatter-style accumulation
- `grad_grid` reduces over channels

The implementation also explicitly aligns with PyTorch for edge behavior:

- non-finite grid coordinates (`nan`, `inf`, `-inf`)
- out-of-bounds coordinates under `zeros`, `border`, and `reflection`
- direct backward interface return semantics for all `output_mask` combinations

## Accuracy Validation

Validation command:

```bash
PYTHONPATH=src python -m pytest tests/test_special_ops.py -k 'grid_sample or grid_sampler' -q
```

Latest result:

```text
588 passed, 6739 deselected
```

Comparison policy:

- Forward 2D/3D: `float16` uses `atol=1e-2`, `float32` uses `atol=1e-4`
- Bicubic special-value cases use `torch.testing.assert_close(..., equal_nan=True)` to match PyTorch `nan` propagation
- Functional backward compares `grad_input` and `grad_grid` against PyTorch autograd
- Direct ATen backward tests compare all four `output_mask` combinations against PyTorch

Coverage summary from `pytest --collect-only`:

| Area | Cases | Notes |
|------|-------|-------|
| 2D forward (`aten.grid_sampler_2d`) | 180 | 5 shapes x 2 dtypes x 3 interpolation modes x 3 padding modes x 2 `align_corners` |
| 3D forward (`aten.grid_sampler_3d`) | 72 | 3 shapes x 2 dtypes x 2 interpolation modes x 3 padding modes x 2 `align_corners` |
| Functional forward (`torch.nn.functional.grid_sample`, 2D) | 18 | Unified functional path coverage |
| Unified dispatch (`aten.grid_sampler`) | 30 | 2D: 18, 3D: 12 |
| Direct backward API (`grid_sampler_{2d,3d}_backward`) | 120 | Covers `output_mask=[TT, TF, FT, FF]` |
| Special values / `nan` / `inf` / `-inf` / out-of-bounds | 90 | 2D: 54, 3D: 36 |
| Empty spatial dimensions | 7 | 2D: 3, 3D: 4 |
| Non-contiguous inputs | 30 | 2D: 18, 3D: 12 |
| Functional backward | 30 | 2D: 18, 3D: 12 |
| Identity / large / error-path checks | 11 | identity grid, large 2D cases, 3D bicubic error path |

## Test Coverage Checklist

| Category | Covered Cases |
|----------|---------------|
| Input scale | small (4x16x16) / medium (16x64x64) / large (64x256x256) |
| Input dims | 4-D (2D grid_sample) / 5-D (3D grid_sample) |
| API branches | forward / backward (grad_input + grad_grid) / functional / direct ATen |
| Interpolation modes | bilinear / nearest / bicubic (2D) / trilinear / nearest (3D) |
| Padding modes | zeros / border / reflection |
| align_corners | True / False |
| DTypes | float16 / float32 |
| Special values | nan / inf / -inf / out-of-bounds coordinates |
| Tensor layout | contiguous / non-contiguous (via .permute()) |
| Empty tensors | (N,C,0,0) / (N,C,H,W)->(N,0,0,2) / (N,C,D,0,H,W) etc. |
| Backward output_mask | [T,T] / [T,F] / [F,T] / [F,F] |
| Error paths | 3D + bicubic raises RuntimeError |

## Benchmark

Benchmark command:

```bash
PYTHONPATH=src python -m pytest benchmark/test_special_perf.py -k 'grid_sampler_2d or grid_sampler_3d' -s --level comprehensive --warmup 20 --iter 50 --record log
```

Latest result:

```text
10 passed, 50 deselected
```

Benchmark source:

- Latest numbers below are taken from [`result_benchmark_test_special_perf-k_grid_sampler_2d or grid_sampler_3d-s--level_comprehensive--warmup_20--iter_50--record_log.log`](/home/FlagGems/result_benchmark_test_special_perf-k_grid_sampler_2d%20or%20grid_sampler_3d-s--level_comprehensive--warmup_20--iter_50--record_log.log)
- Benchmark object is the direct ATen entry (`aten.grid_sampler_2d` / `aten.grid_sampler_3d`), not the functional wrapper
- To keep runtime manageable, the benchmark subset covers `align_corners=False` and two padding modes only: `zeros` / `reflection`

Benchmark shapes from `benchmark/test_special_perf.py`:

- 2D: `(2,16,64,64)->(2,48,48,2)`, `(4,32,128,128)->(4,96,96,2)`, `(1,64,256,256)->(1,192,192,2)`, `(1,64,512,512)->(1,384,384,2)`, `(1,64,768,768)->(1,576,576,2)`
- 3D: `(2,8,16,16,16)->(2,12,12,12,3)`, `(1,32,48,48,48)->(1,36,36,36,3)`, `(2,32,128,128,128)->(2,64,64,64,3)`, `(1,16,256,256,256)->(1,128,128,128,3)`

Measured row count:

- 2D: `3 interpolation modes x 2 padding modes x 5 shapes x 2 dtypes = 60`
- 3D: `2 interpolation modes x 2 padding modes x 4 shapes x 2 dtypes = 32`
- Total: `92` measured rows across `10` benchmark tests

Benchmark summary:

- `87 / 92` measured rows achieve `>= 1.0x`
- 2D: `58 / 60` rows achieve `>= 1.0x`, range `0.926x` to `12.737x`
- 3D: `29 / 32` rows achieve `>= 1.0x`, range `0.962x` to `2.187x`
- The `< 1.0x` rows are isolated cases inside this representative benchmark subset, not a full-mode regression:
  - 2D `bilinear + reflection`, fp32, `(1,64,256,256) -> (1,192,192,2)`: `0.926x`
  - 2D `nearest + zeros`, fp16, `(1,64,256,256) -> (1,192,192,2)`: `0.990x`
  - 3D `trilinear + zeros`, fp32, `(1,32,48,48,48) -> (1,36,36,36,3)`: `0.962x`
  - 3D `trilinear + reflection`, fp16, `(1,32,48,48,48) -> (1,36,36,36,3)`: `0.991x`
  - 3D `trilinear + reflection`, fp32, `(1,32,48,48,48) -> (1,36,36,36,3)`: `0.962x`

Representative benchmark data:

| Dimension | DType | Mode / Padding | Shape | PyTorch (ms) | FlagGems (ms) | Speedup | Status |
|-----------|-------|----------------|-------|--------------|---------------|---------|--------|
| 2D | fp16 | bilinear / zeros | `(2,16,64,64) -> (2,48,48,2)` | 0.017408 | 0.017408 | 1.000x | PASS |
| 2D | fp32 | bilinear / zeros | `(1,64,512,512) -> (1,384,384,2)` | 4.940288 | 1.020928 | 4.839x | PASS |
| 2D | fp16 | bicubic / zeros | `(1,64,768,768) -> (1,576,576,2)` | 15.242240 | 6.621232 | 2.302x | PASS |
| 2D | fp16 | bicubic / reflection | `(2,16,64,64) -> (2,48,48,2)` | 0.247808 | 0.019456 | 12.737x | PASS |
| 2D | fp32 | bilinear / reflection | `(1,64,256,256) -> (1,192,192,2)` | 0.192480 | 0.207872 | 0.926x | BELOW 1.0x |
| 3D | fp16 | trilinear / zeros | `(2,32,128,128,128) -> (2,64,64,64,3)` | 17.958400 | 8.747008 | 2.053x | PASS |
| 3D | fp32 | trilinear / zeros | `(1,32,48,48,48) -> (1,36,36,36,3)` | 0.258048 | 0.268288 | 0.962x | BELOW 1.0x |
| 3D | fp32 | nearest / reflection | `(1,16,256,256,256) -> (1,128,128,128,3)` | 13.390848 | 13.172704 | 1.017x | PASS |

## Checklist Alignment

The current staged files satisfy the competition checklist requirements:

- `src/flag_gems/ops/grid_sample.py`: implementation file present and wired
- `src/flag_gems/ops/__init__.py`: import and `__all__` export updated
- `src/flag_gems/__init__.py`: `_FULL_CONFIG` registration updated, and unified `aten::grid_sampler` autograd path registered
- `tests/test_special_ops.py`: correctness coverage includes forward, functional path, direct backward API, non-finite values, empty tensors, and non-contiguous tensors
- `benchmark/test_special_perf.py`: benchmark subset added with representative medium / large / very large shapes

## Notes

- The direct backward API matches PyTorch default overload behavior for all `output_mask` combinations, including returned optional tensors.
- Reflection and border padding explicitly handle non-finite coordinates to match PyTorch instead of relying on undefined arithmetic on `nan` / `inf`.
- Bicubic mode preserves PyTorch-style `nan` behavior in special-value coverage.
- 2D bicubic achieves high speedups (3x-13x) because PyTorch's CUDA bicubic kernel is not well optimized for the 4x4 sampling pattern, while the Triton kernel fuses all interpolation logic efficiently.

## Files Changed

| File | Description |
|------|-------------|
| `src/flag_gems/ops/grid_sample.py` | grid_sample Triton implementation, dispatch, and backward logic (`2765` lines) |
| `src/flag_gems/ops/__init__.py` | export grid_sample operator symbols |
| `src/flag_gems/__init__.py` | register ATen operators in `_FULL_CONFIG` and register autograd for `aten::grid_sampler` |
| `tests/test_special_ops.py` | add and extend `588` `grid_sample/grid_sampler` tests |
| `benchmark/test_special_perf.py` | add `10` benchmark tests |

### CODE REVIEW

**Comment by goldenfox2025** [comment]
I made a few narrowly scoped fixes to unblock unrelated CI failures on this PR.

Two of them were not in my grid_sample implementation itself:

pad: I fixed a recursion issue in the recent fallback path. constant_pad_nd could loop back into pad, so I separated the core implementation from the dispatch/fallback wrapper to break that cycle.

upsample_bicubic2d_aa_backward: I fixed the test reference path so the PyTorch reference and the tested path use the same quantized input, and I adjusted the quick-CPU reference path to upcast only where CPU does not support fp16/bf16.

I also updated the grid_sample tests to correctly handle quick-CPU reference mode, especially for CPU/GPU device comparison and special NaN/Inf cases.

**Comment by douxetpur** [comment]
Hi, thank you for your participation in the FlagOS Open Computing Competition 🙏

After review, we have chosen to move forward with a different implementation for this operator. As a result, this PR will be closed.

We appreciate your time and contribution, and hope to see more of your submissions in the future.
